java/src/main/java/com/google/appengine/tools/pipeline/impl/util/SerializationUtils.java

```diff
@@ -1,97 +1,60 @@
-    try (ObjectOutputStream out = new ObjectOutputStream(bytes)) {
-      out.writeObject(x);
-    }
-    if (bytes.size() <= MAX_UNCOMPRESSED_BYTE_SIZE) {
-      return bytes.toByteArray();
-    }
-    ByteArrayOutputStream compressedBytes = new ByteArrayOutputStream(bytes.size() / 4);
-    compressedBytes.write(0);
-    compressedBytes.write(ZLIB_COMPRESSION);
-    Deflater deflater =  new Deflater(Deflater.BEST_COMPRESSION, true);
-    try (DeflaterOutputStream out = new DeflaterOutputStream(compressedBytes, deflater)) {
-      // Use internal buffer to avoid copying it.
-      out.write(bytes.getInternalBuffer(), 0, bytes.size());
-    } finally {
-      deflater.end();
+  // err on small; test >= this actually runs faster than < this, suggesting efficiency handling compressed data
+  private static final int MAX_UNCOMPRESSED_BYTE_SIZE = 50_000;
+
+  public static byte[] serialize(Object obj) throws IOException {
+    // Serialize the object
+    ByteArrayOutputStream byteOut = new ByteArrayOutputStream();
+    try (ObjectOutputStream objectOut = new ObjectOutputStream(byteOut)) {
+      objectOut.writeObject(obj);
+    }
+    byte[] serializedData = byteOut.toByteArray();
+
+    // Compress only if serialized data exceeds threshold
+    if (serializedData.length > MAX_UNCOMPRESSED_BYTE_SIZE) {
+      ByteArrayOutputStream compressedByteOut = new ByteArrayOutputStream();
+      try (GZIPOutputStream gzipOut = new GZIPOutputStream(compressedByteOut);
+           ObjectOutputStream objectOut = new ObjectOutputStream(gzipOut)) {
+        objectOut.writeObject(obj);
+        objectOut.flush();
+        gzipOut.finish();
+        serializedData = compressedByteOut.toByteArray();
+      }
     }
-    return compressedBytes.toByteArray();
+    return serializedData;
   }
 
-  public static Object deserialize(byte[] bytes) throws IOException {
-    if (bytes.length < 2) {
-      throw new IOException("Invalid bytes content");
-    }
-    InputStream in = new ByteArrayInputStream(bytes);
-    if (bytes[0] == 0) {
-      in.read(); // consume the marker;
-      if (in.read() != ZLIB_COMPRESSION) {
-        throw new IOException("Unknown compression type");
-      }
-      final Inflater inflater =  new Inflater(true);
-      in = new InflaterInputStream(in, inflater) {
-        @Override public void close() throws IOException {
-          try {
-            super.close();
-          } finally {
-            inflater.end();
-          }
-        }
-      };
-    }
-    try (ObjectInputStream oin = new ObjectInputStream(in)) {
-      try {
-        return oin.readObject();
-      } catch (ClassNotFoundException e) {
-        throw new IOException("Exception while deserilaizing.", e);
+  public static Object deserialize(byte[] data) throws IOException, ClassNotFoundException {
+    // Attempt to decompress
+    try (ByteArrayInputStream byteIn = new ByteArrayInputStream(data);
```

Comment:
could do something like this to avoid the error check
https://github.com/Worklytics/evalengin/blob/c3a95665c4a75a5452cc1aaf8bbed5a44aa4ad1d/worklytics-common/src/main/java/com/engetc/util/CompressionUtils.java#L41